tinygrad/shape/shapetracker.py

```diff
@@ -149,85 +200,72 @@ def expr_idxs(self, idxs=None):
     return self._expr_idx(idx, valid)
 
   def expr_node(self, idx='idx'):
-    if isinstance(idx, str): idx = Variable(idx, 0, prod(self.shape)-1)
+    if idx.__class__ == str: idx = Variable(idx, 0, prod(self.shape)-1)
     return self._expr_idx(self.views[-1].expr_node(idx), self.views[-1].expr_node_mask(idx))
 
   def needs_valid(self) -> bool:
-    return any(v.mask is not None for v in self.views)
+    return any([v.mask is not None for v in self.views])
 
   # *** under this line are the movement ops ***
 
   def __unsafe_resize(self, arg: Tuple[Tuple[int, int], ...], mask=None):
-    offset = sum([self.views[-1].strides[i]*x for i,(x,_) in enumerate(arg)])
+    offset = get_unsafe_resize_offset(self.views[-1].strides, arg)
     if self.views[-1].mask:
       # move the old mask
-      nmask = tuple((max(mx-ax, 0), min(my-ax, ay-ax)) for (mx,my),(ax,ay) in zip(self.views[-1].mask, arg))
+      nmask = tuple([(max(mx-ax, 0), min(my-ax, ay-ax)) for (mx,my),(ax,ay) in zip(self.views[-1].mask, arg)])
       # merge the masks if we have two
-      mask = tuple((max(mx1, mx2), min(my1, my2)) for (mx1, my1), (mx2, my2) in zip(nmask, mask)) if mask is not None else nmask
-    self.views[-1] = View(tuple(y-x for x,y in arg), self.views[-1].strides, self.views[-1].offset+offset, mask)
+      mask = tuple([(max(mx1, mx2), min(my1, my2)) for (mx1, my1), (mx2, my2) in zip(nmask, mask)]) if mask is not None else nmask
+    self.views[-1] = View(tuple([y-x for x,y in arg]), self.views[-1].strides, self.views[-1].offset+offset, mask)
 
   def pad(self, arg: Tuple[Tuple[int, int], ...]):
     assert all((b>=0 and e>=0) for b,e in arg) and len(arg) == len(self.shape)
-    if all(b==0 and e==0 for b,e in arg): return self
-    zvarg = tuple((-b,s+e) for s,(b,e) in zip(self.shape, arg))
-    self.__unsafe_resize(zvarg, mask=tuple((b,s+b) for s,(b,_) in zip(self.shape, arg)))
+    if any([b or e for b, e in arg]):
+      zvarg, mask = get_pad_args(self.shape, arg)
+      self.__unsafe_resize(zvarg, mask=mask)
+      return self
+    return self
```

Comment:
Why do these functions have to `return self`? Isn't this dealt with in the `movement_op` dispatcher?

Reply:
Artifact of the optimizations in `lazy.py` in the big PR, the functions return self because we later get rid of the dispatcher altogether.

I can revert that for this PR to keep things more contained